tests: kernel: mutex: move race timeout test to mutex_api

cfriedt opened the pull request:
Previously, this change was added to `mutex_error_case`.

That worked fine in `main`, but once the change was backported to `v2.7-branch`, the test would fail because it *did not* cause a failure. The reason for that, was that the `mutex_error_case` suite has `CONFIG_ZTEST_FATAL_HOOK=y`.

With the newer ztest API, it allowed a separate suite to be used, allowing the test to pass (although it did not really fit in with the rest of the testsuite).

The solution is to simply merge it with the `mutex_api` suite which uses non-inverted success logic.

This change will also have to be cherry-picked for the backport in #49031.

Fixes #48056.

cc @yperess - not sure if this exposes any weird state in the new testsuite

cfriedt commented:
Just a thought - would it be better here to revert the previous commit and simply use this one instead? That way there is less to backport.

peter-mitsis commented on tests/kernel/mutex/mutex_api/src/test_mutex_apis.c:
```diff
@@ -387,6 +387,52 @@ ZTEST_USER(mutex_api_1cpu, test_mutex_priority_inheritance)
+			tThread_mutex_lock_should_fail, &mutex, NULL, NULL, 8,
+			K_USER | K_INHERIT_PERMS, K_NO_WAIT);
+
+	k_mutex_lock(&mutex, K_FOREVER);
```
Minor item: I would prefer to see that *k_mutex_lock()* is moved to before *k_thread_create()* is called. This makes it blatantly obvious to a reader that the mutex is locked before the helper thread executes. Otherwise anyone reading the code in the future has to expend additional energy/time verifying priorities, scheduling, and number of cpus involved to ensure that  it behaves as described.

peter-mitsis replied:
Back up a moment ... isn't the priority of the running thread here a higher priority level than the created thread? If true, then the current thread gets the thread and when the created thread tries to take the semaphore, it would not lead to a case of priority inversion.

Thus, we should be doing ...

1. lock mutex
2. create helper thread at higher priority to current thread.

cfriedt commented:
Since I'm responding this anyway to address @peter-mitsis suggestions, I think I will opt to revert the previous commit that add d the test in the wrong suite, because then a single Testcase can be backported a bit more cleanly.

@peter-mitsis @andyross - is that OK with both of you?

peter-mitsis commented on tests/kernel/mutex/mutex_api/src/test_mutex_apis.c:
```diff
@@ -387,6 +387,52 @@ ZTEST_USER(mutex_api_1cpu, test_mutex_priority_inheritance)
+
+	k_mutex_lock(&mutex, K_FOREVER);
+
+	k_sleep(K_MSEC(TIMEOUT));
```
If priority inheritance is kicking in, then both the helper and the current thread should be at the same priority. They are both likely to wake on the same tick. The test as written is assuming that the current thread is going to be executed first. Such an assumption might not hold--it is highly dependent upon the implementation of the scheduler and which thread is got to the "head" of the ready queue first.

To work around this, I have two ideas...
1. If available, use an absolute timeout for both the helper and current thread to guarantee that both threads will expire on the same tick. 
2. Before the current thread sleeps, explicitly bump its priority even higher. This should help guarantee that it will always execute before the helper thread.

cfriedt replied:
@peter-mitsis - I could create an arbitrary absolute tick value in the current thread and pass that as a parameter to the helper thread.

One option is making that a Kconfig value local to the testsuite.

The main concern I have with that is that it could be another one of those racey values that we just have to keep increasing over time. Is there a happy medium between long-enough to allow the thread to start and short enough to not be considered "too long for a test"?

peter-mitsis commented:
@cfriedt - I'm good with your proposal.

Commit:
Revert "tests: kernel: mutex: test for lock timeout race"

This reverts commit 019a6ecae352ee6a5f7d2c1eb15c282f0858cc6d.

Signed-off-by: Chris Friedt <cfriedt@meta.com>

Commit:
tests: kernel: mutex: move race timeout test to mutex_api

Previously, this change was added to `mutex_error_case`.

That worked fine in `main`, but once the change was backported to
`v2.7-branch`, the test would fail because it *did not* cause a
failure. The reason for that, was that the `mutex_error_case`
suite has `CONFIG_ZTEST_FATAL_HOOK=y`.

With the newer ztest API, it allowed a separate suite to be used,
allowing the test to pass (although it did not really fit in with
the rest of the testsuite).

The solution is to simply merge it with the `mutex_api` suite
which uses non-inverted success logic.

This change will also have to be cherry-picked for the backport
in #49031.

Fixes #48056.

tests: kernel: mutex: move race timeout test to mutex_api

Previously, this change was added to `mutex_error_case`.

That worked fine in `main`, but once the change was backported to
`v2.7-branch`, the test would fail because it *did not* cause a
failure. The reason for that, was that the `mutex_error_case`
suite has `CONFIG_ZTEST_FATAL_HOOK=y`.

With the newer ztest API, it allowed a separate suite to be used,
allowing the test to pass (although it did not really fit in with
the rest of the testsuite).

The solution is to simply merge it with the `mutex_api` suite
which uses non-inverted success logic.

This change will also have to be cherry-picked for the backport
in #49031.

Fixes #48056.

Signed-off-by: Chris Friedt <cfriedt@meta.com>

cfriedt commented:
@peter-mitsis, @dcpleung - please have another look. I believe this addresses all remaining concerns.

cfriedt commented:
@peter-mitsis - can you approve?
Maybe @andyross ?

andyross reviewed (approved):
Makes sense.  Could maybe consider merging all those error_case tests too?

cfriedt commented:
> Makes sense.  Could maybe consider merging all those error_case tests too?

Makes sense - I'll check to see what is missing.
